client/src/components/Tool/ToolForm.vue

```diff
@@ -0,0 +1,300 @@
+                return "The previous run of this tool failed. Use this option to replace the failed element(s) in the dataset collection that were produced during the previous tool run.";
+            } else {
+                return "The previous run of this tool failed and other tools were waiting for it to finish successfully. Use this option to resume those tools using the new output(s) of this tool run.";
+            }
+        },
+        reuseAllowed() {
+            const Galaxy = getGalaxyInstance();
+            if (Galaxy.user.attributes.preferences && "extra_user_preferences" in Galaxy.user.attributes.preferences) {
+                const extra_user_preferences = JSON.parse(Galaxy.user.attributes.preferences.extra_user_preferences);
+                const keyCached = "use_cached_job|use_cached_job_checkbox";
+                const hasCachedJobs = keyCached in extra_user_preferences;
+                return hasCachedJobs ? extra_user_preferences[keyCached] : false;
+            } else {
+                return false;
+            }
+        },
+    },
+    methods: {
+        onValidation(validationInternal) {
+            this.validationInternal = validationInternal;
+        },
+        onChange(newData) {
+            this.formData = newData;
+            this.onUpdate();
+        },
+        onUpdate() {
+            updateTool(this.formConfig.id, this.currentVersion, this.formData).then((data) => {
+                this.formConfig = data;
+            });
+        },
+        onChangeVersion(newVersion) {
+            this.requestTool(newVersion);
+        },
+        onChangeEmail(email) {
+            this.email = email;
+        },
+        onChangeRemap(remap) {
+            this.remap = remap ? this.job_id : null;
+        },
+        onChangeReuse(reuse) {
+            this.reuse = reuse;
+        },
+        onUpdateFavorites(user, newFavorites) {
+            user.preferences["favorites"] = newFavorites;
+        },
+        requestTool(newVersion) {
+            this.currentVersion = newVersion || this.currentVersion;
+            getTool(this.id, this.currentVersion, this.job_id).then((data) => {
+                this.formConfig = data;
+                this.showLoading = false;
+                this.showForm = true;
+                if (newVersion) {
+                    this.messageVariant = "success";
+                    this.messageText = `Now you are using '${data.name}' version ${data.version}, id '${data.id}'.`;
+                }
+            });
+        },
+        onExecute() {
+            if (this.validationInternal) {
+                this.validationScrollTo = this.validationInternal.slice();
+                return;
+            }
+            this.showExecuting = true;
+            const options = this.formData;
+            const Galaxy = getGalaxyInstance();
+            const history_id = Galaxy.currHistoryPanel && Galaxy.currHistoryPanel.model.id;
```

Comment:
Not sure that's the right way to get the current history id, and it should probably be consistent with what the tool form had been built with ?

Reply:
The real problem is that it is in a method which must be manually called. The value will not be dynamic as the currentHistory changes from outside this component. History should probably be a property of the component and its value should be pulled from Vuex, or a provider.

Reply:
Not sure what you mean with "save" function but overall we are on the same page regarding the goal towards we are moving. Is there a history provider I can use, or should I add a preliminary one within this PR?

Reply:
There is the UserHistories provider, but maybe the watches in there aren't right for this situation. I guess a separate CurrentHistoryId provider might make sense ? It can take the value from the betaHistory vuex store (https://github.com/galaxyproject/galaxy/blob/11363a013ee187baaa0e7733ed909dfa791f38be/client/src/components/History/model/historyStore.js#L46)

Reply:
@mvdbeek 
 I see two watches in UserHistories, one which reloads the list of histories if the user changes and one which reloads the history data if the user selects a new history. Is one or the other of these not behaving well?

@guerler 
I was just suggesting that if the form emitted a "complete" event when the data was ready, then you could modularly separate out the components whose job is to get the big object of data (the form components) and whatever server-side interaction. I was using "save" as a generic term to indicate "whatever you do with the form data". Ex:

```vue
<template>
    <!-- I am some interface which you have decided could use a toolInputform 
    to generate a big object of data that will be used in some other process

     Here I've broken v-model into value/@input to show how you could decide to use the returning
     data in any number of ways if the end result were presented as a Vue event, even if just as the built-in
     event name that comes baked-in with v-model.  With components, it is the use of events that makes
     separation of concerns possible. It will make your components more usable, testable, and flexible. -->
    <ToolForm :config="formConfiguration" :value="inputs" @input="thisIsTheSaveFunction" />
</template>

<script>
export default {
    methods: {
        thisIsTheSaveFunction(valFromForm) {
               // You could save the value to the server, start a tool running, update something in the store, whatever you
               // need to do. The point is to decouple the data retrieval from whatever you do with the data
         }
    }
}
</script>
```

Reply:
This is pretty much how it looks. In the last stage, we can revisit all locations in which the `FormDisplay` component is used and we can use `input` event instead of the `change` event and further smooth out the component interface.